### PR TITLE
feat(spark): integrate Spark operator in Kubeflow manifests

### PR DESCRIPTION
**Pull request by GezimSejdiu**
This PR introduces the integration of the Spark operator into the Kubeflow manifests, enabling users to run distributed Spark workloads directly within their Kubeflow environment. With this addition, Kubeflow users can leverage Spark for large-scale data processing as part of their machine learning (ML) pipelines, improving scalability and efficiency.

#### **Key Changes**:
- Added Spark operator manifests to support Spark job execution.
- Ensured integration of the Spark operator with Kubeflow pipelines, allowing users to submit and monitor Spark jobs within the Kubeflow UI.
- Updated the documentation to include instructions on how to deploy and use the Spark operator within Kubeflow.
- Configured role-based access control (RBAC) to secure Spark job submissions and resource usage.
  
#### **Benefits**:
- Enables distributed data processing using Spark directly within Kubeflow.
- Simplifies the management and orchestration of Spark jobs as part of ML workflows.
- Provides an easy-to-use interface to monitor and manage Spark workloads.

#### **How to Test**:
1. Deploy the updated Kubeflow manifests, including the Spark operator.
2. Submit a sample Spark job via the Kubeflow pipeline UI.
3. Verify the job execution, logs, and results within the Spark UI and Kubeflow pipeline dashboard.

### CODE REVIEW

**Review by andreyvelich — COMMENTED** [review]
Thank you so much for doing this @GezimSejdiu!
Given that Spark Operator is actually part of Kubeflow core components, we should install it as other Kubeflow apps: https://github.com/kubeflow/manifests/tree/master/apps.

/assign @ChenYi015  @yuchaoran2011 @vara-bonthu @ImpSy @jacobsalway

**Comment by juliusvonkohout** [comment]
@andreyvelich  @GezimSejdiu is a colleague of mine ;-) Yes, we can add it to applications directly since you seem to be fine with it as owner of spark-operator https://github.com/kubeflow/spark-operator/blob/3acd0f1a900a933e8612c1b4af55d29b1112cbf1/OWNERS#L2

File: contrib/spark/test.sh
Comment (juliusvonkohout) [inline contrib/spark/test.sh:1]: Please use https://github.com/kubeflow/manifests/blob/master/.github/workflows/ray_test.yaml as base such that we really test a Kubeflow user namespace.

**Comment by GezimSejdiu** [comment]
> Thank you so much for doing this @GezimSejdiu! Given that Spark Operator is actually part of Kubeflow core components, we should install it as other Kubeflow apps: https://github.com/kubeflow/manifests/tree/master/apps.
> 
> /assign @ChenYi015 @yuchaoran2011 @vara-bonthu @ImpSy @jacobsalway

Hey @andreyvelich , 

thank you. Good point. While thinking now, I still believe we can keep it into contrib in case, at some point we decide to merge efforts with Apache Spark Operator, but let us discuss that in another thread. @juliusvonkohout what do you think? Shall we move them into apps or still consider keeping (for now) under contrib.

**Comment by GezimSejdiu** [comment]
I'm currently testing it locally and will be working on it to make it work, but it seems webhook isn't working properly, I get: 
```bash
2024-10-14T11:17:20.496Z	ERROR	webhook/start.go:254	Failed to sync webhook secret	{"error": "secrets \"spark-operator-webhook-certs\" is forbidden: User \"system:serviceaccount:kubeflow:spark-operator-webhook\" cannot get resource \"secrets\" in API group \"\" in the namespace \"default\""}
github.com/kubeflow/spark-operator/cmd/operator/webhook.start
	/workspace/cmd/operator/webhook/start.go:254
github.com/kubeflow/spark-operator/cmd/operator/webhook.NewStartCommand.func2
	/workspace/cmd/operator/webhook/start.go:128
github.com/spf13/cobra.(*Command).execute
	/go/pkg/mod/github.com/spf13/cobra@v1.8.1/command.go:989
github.com/spf13/cobra.(*Command).ExecuteC
	/go/pkg/mod/github.com/spf13/cobra@v1.8.1/command.go:1117
github.com/spf13/cobra.(*Command).Execute
	/go/pkg/mod/github.com/spf13/cobra@v1.8.1/command.go:1041
main.main
	/workspace/cmd/main.go:27
runtime.main
	/usr/local/go/src/runtime/proc.go:272

```

I think we will have to play with webhook configurations to make this work. 

@andreyvelich  feel free to suggest any solution for this issue.

**Comment by andreyvelich** [comment]
> > Thank you so much for doing this @GezimSejdiu! Given that Spark Operator is actually part of Kubeflow core components, we should install it as other Kubeflow apps: https://github.com/kubeflow/manifests/tree/master/apps.
> > /assign @ChenYi015 @yuchaoran2011 @vara-bonthu @ImpSy @jacobsalway
> 
> Hey @andreyvelich ,
> 
> thank you. Good point. While thinking now, I still believe we can keep it into contrib in case, at some point we decide to merge efforts with Apache Spark Operator, but let us discuss that in another thread. @juliusvonkohout what do you think? Shall we move them into apps or still consider keeping (for now) under contrib.

Given that Kubeflow community currently maintain this operator, I think we should just move it to apps since it is part of Kubeflow core components as described here: https://www.kubeflow.org/docs/started/architecture/#kubeflow-ecosystem
Furthermore, we have plans to release Kubeflow Spark Operator as part of Kubeflow 1.10 release.

Any concerns with that @ChenYi015 @yuchaoran2011 @vara-bonthu @ImpSy @jacobsalway?

**Comment by juliusvonkohout** [comment]
> I'm currently testing it locally and will be working on it to make it work, but it seems webhook isn't working properly, I get:
> 
> ```shell
> 2024-10-14T11:17:20.496Z	ERROR	webhook/start.go:254	Failed to sync webhook secret	{"error": "secrets \"spark-operator-webhook-certs\" is forbidden: User \"system:serviceaccount:kubeflow:spark-operator-webhook\" cannot get resource \"secrets\" in API group \"\" in the namespace \"default\""}
> github.com/kubeflow/spark-operator/cmd/operator/webhook.start
> 	/workspace/cmd/operator/webhook/start.go:254
> github.com/kubeflow/spark-operator/cmd/operator/webhook.NewStartCommand.func2
> 	/workspace/cmd/operator/webhook/start.go:128
> github.com/spf13/cobra.(*Command).execute
> 	/go/pkg/mod/github.com/spf13/cobra@v1.8.1/command.go:989
> github.com/spf13/cobra.(*Command).ExecuteC
> 	/go/pkg/mod/github.com/spf13/cobra@v1.8.1/command.go:1117
> github.com/spf13/cobra.(*Command).Execute
> 	/go/pkg/mod/github.com/spf13/cobra@v1.8.1/command.go:1041
> main.main
> 	/workspace/cmd/main.go:27
> runtime.main
> 	/usr/local/go/src/runtime/proc.go:272
> ```
> 
> I think we will have to play with webhook configurations to make this work.
> 
> @andreyvelich feel free to suggest any solution for this issue.

You need a networkpolicy as in https://github.com/kubeflow/manifests/blob/master/common/networkpolicies/base/training-operator-webhook.yaml

File: contrib/spark/spark-operator/base/kustomization.yaml [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+patches:
+  # Add securityContext to Spark Operator Pod.
+  - target:
+      kind: Deployment
+      labelSelector: "app.kubernetes.io/name=spark-operator"
+    patch: |-
+      - op: add
+        path: /spec/template/spec/containers/0/securityContext
+        value:
+          runAsUser: 1000
```
Comment (ImpSy) [inline contrib/spark/spark-operator/base/kustomization.yaml:16]: The operator is based on spark image that uses 185 as non root user

Comment (GezimSejdiu) [reply contrib/spark/spark-operator/base/kustomization.yaml:16]: Good catch. Let me then remove this as we have the non-root user by default on Spark.

Comment (juliusvonkohout) [reply contrib/spark/spark-operator/base/kustomization.yaml:16]: Just runAsNonRoot:true is enough then

**Comment by GezimSejdiu** [comment]
> I see for example https://github.com/kubeflow/manifests/blob/a6f16d41230b43a375a2f95478519303ae02eef8/contrib/spark/spark-operator/base/resources.yaml#L23191 and many other places with the namespace default. It needs to be adjusted to run in the namespace "kubeflow." Maybe you can just do so in the main kustomization.yaml with a single line.

Indeed, that was the issue. I made it work by specifying the namespace while generating the resources.yaml and also removing the default namespace from spark.jobNamesapces (which is the default on the helm chart) to restrict to only that namespace. 

I made it work now, but only without webhook. When I enable webhook I get this: 
```bash
Error from server (InternalError): error when creating "sparkapplication_example.yaml": Internal error occurred: failed calling webhook "mutate-sparkoperator-k8s-io-v1beta2-sparkapplication.sparkoperator.k8s.io": failed to call webhook: Post "https://spark-operator-webhook-svc.kubeflow.svc:9443/mutate-sparkoperator-k8s-io-v1beta2-sparkapplication?timeout=10s": dial tcp 10.96.184.97:9443: i/o timeout
```
and when I set `webhook.port=443` it gives me: 
```
Error from server (InternalError): error when creating "sparkapplication_example.yaml": Internal error occurred: failed calling webhook "mutate-sparkoperator-k8s-io-v1beta2-sparkapplication.sparkoperator.k8s.io": failed to call webhook: Post "https://spark-operator-webhook-svc.kubeflow.svc:443/mutate-sparkoperator-k8s-io-v1beta2-sparkapplication?timeout=10s": dial tcp 10.96.125.227:443: connect: connection refused

```

I saw that you override the service e.g. https://github.com/kubeflow/manifests/blob/master/apps/tensorboard/tensorboard-controller/upstream/webhook/service.yaml and I did try to do the same, but it doesn't seem to work.

Shall we consider disabling webhook or it is a must and we have to enforce such validation. If we can go ahead, I think this version is working already, I did test it locally. Let us also see / test CICD.

File: contrib/spark/README.md [hunk 2]
```diff
@@ -0,0 +1,26 @@
+# Kubeflow Spark Operator
```
Comment (andreyvelich) [inline contrib/spark/README.md:1]: If you move it to apps, you don't need to have README like for other Kubeflow components: https://github.com/kubeflow/manifests/tree/master/apps
We already have all of the required docs here: https://www.kubeflow.org/docs/components/spark-operator/

Comment (juliusvonkohout) [reply contrib/spark/README.md:1]: We can follow up in another PR. This is still an experiment.

**Comment by juliusvonkohout** [comment]
/lgtm
/approve

We might continue in follow up PRs.

**Comment by andreyvelich** [comment]
@GezimSejdiu @juliusvonkohout Can you create tracking issue to move Spark Operator into Kubeflow apps from contrib folder ?

**Comment by andreyvelich** [comment]
What is the motivation to install apache-spark-operator, if right now Kubeflow community maintains Kubeflow Spark Operator and contributes features there ?

**Comment by andreyvelich** [comment]
Additionally, the Kubeflow Spark Operator will be included in Kubeflow 1.10 release which means it should be part of Kubeflow apps in the manifest repo.
cc @rimolive

**Comment by juliusvonkohout** [comment]
> What is the motivation to install apache-spark-operator, if right now Kubeflow community maintains Kubeflow Spark Operator and contributes features there ?
> 

We are just not sure whether it also supports interactive spark sessions live from a jupyterlab and persistent clusters next to just running packaged spark applications.

**Comment by vikas-saxena02** [comment]
@GezimSejdiu are you implementing the support with Kubeflow Pipelines v1 or v2 or both?

**Comment by vikas-saxena02** [comment]
> > What is the motivation to install apache-spark-operator, if right now Kubeflow community maintains Kubeflow Spark Operator and contributes features there ?
> 
> We are just not sure whether it also supports interactive spark sessions live from a jupyterlab and persistent clusters next to just running packaged spark applications.

@juliusvonkohout I have been working on implementing the interactive support but seems like this will be a big one and will need support for Spark Connect to be implemented in SparkOperator first. This may take  sometime but is definitely work in progress.  Also, interactive spark jobs work well with jupyter enterprise gateway or Apache Levy, the proposal doc has this updated.

**Pull request merged**